# Task-4.8

# Step 1 Import the relevant libraries, along with your ords_prods_merge dataframe

In [3]:
# Import Libraries

import pandas as pd
import numpy as np
import os

In [5]:
# setting path
path = r'/Users/shikhakamboj/Documents/JAN-2025 Instacart Basket Analysis'

# import data sets
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [7]:
ords_prods_merge.shape

(32404859, 18)

# Step 2 Aggregate Mean Calculation for Entire Dataframe

In [12]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3 Analyze the result. How do the results for the entire dataframe differ from those of the subset?
Comparison Between Subset and Entire DataFrame:
 1. Highest Average:  Entire Dataset: Department 21 = 22.90, Subset: Department 21 = 23.26
 2. Significant Differences:  Department 10: Entire = 20.19 and Subset = 21.29, Department 2: Entire = 17.28 and Subset = 18.46
 3. Departments 6, 7, 13, 14, 15 show minimal variations between the two datasets.

# Step 4 Creating a loyalty flag for existing customers using the transform() and loc() functions.

In [19]:
# Aggregating Data with transform()

ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform('max')

In [25]:
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0          2   202279             3                   5                  9   
1          2   202279             3                   5                  9   
2          2   202279             3                   5                  9   
3          2   202279             3                   5                  9   
4          2   202279             3                   5                  9   
5          2   202279             3                   5                  9   
6          2   202279             3                   5                  9   
7          2   202279             3                   5                  9   
8          2   202279             3                   5                  9   
9          3   205970            16                   5                 17   
10         3   205970            16                   5                 17   
11         3   205970            16                   5                 17   
12         3   205970            16                   5                 17   
13         3   205970            16                   5                 17   
14         3   205970            16                   5                 17   
15         3   205970            16                   5                 17   
16         3   205970            16                   5                 17   
17         4   178520            36                   1                  9   
18         4   178520            36                   1                  9   
19         4   178520            36                   1                  9   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      8.0     33120.0                1.0        1.0   
1                      8.0     28985.0                2.0        1.0   
2                      8.0      9327.0                3.0        0.0   
3                      8.0     45918.0                4.0        1.0   
4                      8.0     30035.0                5.0        0.0   
5                      8.0     17794.0                6.0        1.0   
6                      8.0     40141.0                7.0        1.0   
7                      8.0      1819.0                8.0        1.0   
8                      8.0     43668.0                9.0        0.0   
9                     12.0     33754.0                1.0        1.0   
10                    12.0     24838.0                2.0        1.0   
11                    12.0     17704.0                3.0        1.0   
12                    12.0     21903.0                4.0        1.0   
13                    12.0     17668.0                5.0        1.0   
14                    12.0     46667.0                6.0        1.0   
15                    12.0     17461.0                7.0        1.0   
16                    12.0     32665.0                8.0        1.0   
17                     7.0     46842.0                1.0        0.0   
18                     7.0     26434.0                2.0        1.0   
19                     7.0     39758.0                3.0        1.0   

                                         product_name  aisle_id  \
0                                  Organic Egg Whites        86   
1                               Michigan Organic Kale        83   
2                                       Garlic Powder       104   
3                                      Coconut Butter        19   
4                                   Natural Sweetener        17   
5                                             Carrots        83   
6                    Original Unflavored Gelatine Mix       105   
7            All Natural No Stir Creamy Almond Butter        88   
8                             Classic Blend Cole Slaw       123   
9   Total 2% with Strawberry Lowfat Greek Strained...       120   
10                             Unsweetened Almondmilk        91   
11                                             L

In [27]:
# Deriving Columns with loc()

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [29]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [31]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [33]:
ords_prods_merge['loyalty_flag'].value_counts(dropna= False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [35]:
# updated output

ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0          2   202279             3                   5                  9   
1          2   202279             3                   5                  9   
2          2   202279             3                   5                  9   
3          2   202279             3                   5                  9   
4          2   202279             3                   5                  9   
5          2   202279             3                   5                  9   
6          2   202279             3                   5                  9   
7          2   202279             3                   5                  9   
8          2   202279             3                   5                  9   
9          3   205970            16                   5                 17   
10         3   205970            16                   5                 17   
11         3   205970            16                   5                 17   
12         3   205970            16                   5                 17   
13         3   205970            16                   5                 17   
14         3   205970            16                   5                 17   
15         3   205970            16                   5                 17   
16         3   205970            16                   5                 17   
17         4   178520            36                   1                  9   
18         4   178520            36                   1                  9   
19         4   178520            36                   1                  9   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      8.0     33120.0                1.0        1.0   
1                      8.0     28985.0                2.0        1.0   
2                      8.0      9327.0                3.0        0.0   
3                      8.0     45918.0                4.0        1.0   
4                      8.0     30035.0                5.0        0.0   
5                      8.0     17794.0                6.0        1.0   
6                      8.0     40141.0                7.0        1.0   
7                      8.0      1819.0                8.0        1.0   
8                      8.0     43668.0                9.0        0.0   
9                     12.0     33754.0                1.0        1.0   
10                    12.0     24838.0                2.0        1.0   
11                    12.0     17704.0                3.0        1.0   
12                    12.0     21903.0                4.0        1.0   
13                    12.0     17668.0                5.0        1.0   
14                    12.0     46667.0                6.0        1.0   
15                    12.0     17461.0                7.0        1.0   
16                    12.0     32665.0                8.0        1.0   
17                     7.0     46842.0                1.0        0.0   
18                     7.0     26434.0                2.0        1.0   
19                     7.0     39758.0                3.0        1.0   

                                         product_name  aisle_id  \
0                                  Organic Egg Whites        86   
1                               Michigan Organic Kale        83   
2                                       Garlic Powder       104   
3                                      Coconut Butter        19   
4                                   Natural Sweetener        17   
5                                             Carrots        83   
6                    Original Unflavored Gelatine Mix       105   
7            All Natural No Stir Creamy Almond Butter        88   
8                             Classic Blend Cole Slaw       123   
9   Total 2% with Strawberry Lowfat Greek Strained...       120   
10                             Unsweetened Almondmilk        91   
11                                             L

# Step 5  Analyze whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [38]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

New customers tend to spend more on average per purchase, while loyal customers spend the least. All groups share the same minimum (1.0) and maximum (99,999.0) price points.

# Step 6 Create a spending flag for each user based on the average price across all their orders using given criteria.

In [42]:
# Aggregating Data with transform()

ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [48]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0         2   202279             3                   5                  9   
1         2   202279             3                   5                  9   
2         2   202279             3                   5                  9   
3         2   202279             3                   5                  9   
4         2   202279             3                   5                  9   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     8.0     33120.0                1.0        1.0   
1                     8.0     28985.0                2.0        1.0   
2                     8.0      9327.0                3.0        0.0   
3                     8.0     45918.0                4.0        1.0   
4                     8.0     30035.0                5.0        0.0   

            product_name  ...  department_id  prices  match  \
0     Organic Egg Whites  ...             16    11.3   both   
1  Michigan Organic Kale  ...              4    13.4   both   
2          Garlic Powder  ...             13     3.6   both   
3         Coconut Butter  ...             13     8.4   both   
4      Natural Sweetener  ...             13    13.7   both   

     price_range_loc     busiest day    busiest days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most orders   
1  Mid-range product  Regularly busy  Regularly busy           Most orders   
2  Low-range product  Regularly busy  Regularly busy           Most orders   
3  Mid-range product  Regularly busy  Regularly busy           Most orders   
4  Mid-range product  Regularly busy  Regularly busy           Most orders   

  max_order  loyalty_flag avg_product_price  
0         8  New customer          8.618889  
1         8  New customer          8.618889  
2         8  New customer          8.618889  
3         8  New customer          8.618889  
4         8  New customer          8.618889  

[5 rows x 21 columns]

In [50]:
# Deriving "Spending_flag "Columns with loc()

ords_prods_merge.loc[ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [52]:
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'

In [54]:
ords_prods_merge['spending_flag'].value_counts(dropna= False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [64]:
ords_prods_merge[['user_id', 'spending_flag', 'avg_product_price']].head(10)

user_id spending_flag  avg_product_price
0   202279   Low spender           8.618889
1   202279   Low spender           8.618889
2   202279   Low spender           8.618889
3   202279   Low spender           8.618889
4   202279   Low spender           8.618889
5   202279   Low spender           8.618889
6   202279   Low spender           8.618889
7   202279   Low spender           8.618889
8   202279   Low spender           8.618889
9   205970   Low spender           8.681734

# Step 7 Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

In [69]:
# Aggregating Data with transform()

ords_prods_merge['user_ordering_behavior'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [71]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0         2   202279             3                   5                  9   
1         2   202279             3                   5                  9   
2         2   202279             3                   5                  9   
3         2   202279             3                   5                  9   
4         2   202279             3                   5                  9   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     8.0     33120.0                1.0        1.0   
1                     8.0     28985.0                2.0        1.0   
2                     8.0      9327.0                3.0        0.0   
3                     8.0     45918.0                4.0        1.0   
4                     8.0     30035.0                5.0        0.0   

            product_name  ...  match    price_range_loc     busiest day  \
0     Organic Egg Whites  ...   both  Mid-range product  Regularly busy   
1  Michigan Organic Kale  ...   both  Mid-range product  Regularly busy   
2          Garlic Powder  ...   both  Low-range product  Regularly busy   
3         Coconut Butter  ...   both  Mid-range product  Regularly busy   
4      Natural Sweetener  ...   both  Mid-range product  Regularly busy   

     busiest days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy           Most orders         8  New customer   
1  Regularly busy           Most orders         8  New customer   
2  Regularly busy           Most orders         8  New customer   
3  Regularly busy           Most orders         8  New customer   
4  Regularly busy           Most orders         8  New customer   

  avg_product_price  spending_flag user_ordering_behavior  
0          8.618889    Low spender                   24.0  
1          8.618889    Low spender                   24.0  
2          8.618889    Low spender                   24.0  
3          8.618889    Low spender                   24.0  
4          8.618889    Low spender                   24.0  

[5 rows x 23 columns]

In [73]:
# Deriving "frequency_flag "Columns with loc()

ords_prods_merge.loc[ords_prods_merge['user_ordering_behavior'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [81]:
ords_prods_merge.loc[(ords_prods_merge['user_ordering_behavior'] <= 20) & (ords_prods_merge['user_ordering_behavior'] > 10), 'frequency_flag'] = 'Regular customer'

In [83]:
ords_prods_merge.loc[ords_prods_merge['user_ordering_behavior'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [85]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [89]:
# Checking Relevant Columns of Updated Dataframe

ords_prods_merge[['user_id', 'days_since_prior_order', 'user_ordering_behavior', 'frequency_flag']].head(20)

user_id  days_since_prior_order  user_ordering_behavior  \
0    202279                     8.0                    24.0   
1    202279                     8.0                    24.0   
2    202279                     8.0                    24.0   
3    202279                     8.0                    24.0   
4    202279                     8.0                    24.0   
5    202279                     8.0                    24.0   
6    202279                     8.0                    24.0   
7    202279                     8.0                    24.0   
8    202279                     8.0                    24.0   
9    205970                    12.0                     8.0   
10   205970                    12.0                     8.0   
11   205970                    12.0                     8.0   
12   205970                    12.0                     8.0   
13   205970                    12.0                     8.0   
14   205970                    12.0                     8.0   
15   205970                    12.0                     8.0   
16   205970                    12.0                     8.0   
17   178520                     7.0                     7.0   
18   178520                     7.0                     7.0   
19   178520                     7.0                     7.0   

           frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
5   Non-frequent customer  
6   Non-frequent customer  
7   Non-frequent customer  
8   Non-frequent customer  
9       Frequent customer  
10      Frequent customer  
11      Frequent customer  
12      Frequent customer  
13      Frequent customer  
14      Frequent customer  
15      Frequent customer  
16      Frequent customer  
17      Frequent customer  
18      Frequent customer  
19      Frequent customer

# Step 9 Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [92]:
# Exporting the Dataframe

ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_grouped_aggregated.pkl'))